src/mito2/src/compaction/memory_manager.rs

```diff
@@ -0,0 +1,711 @@
+                }
+
+                let permit = inner.semaphore.acquire_many(permits).await.unwrap();
+                permit.forget();
+                inner.on_acquire();
+                CompactionMemoryGuard::limited(inner.clone(), permits)
+            }
+        }
+    }
+
+    /// Tries to acquire memory. Returns Some(guard) on success, None if insufficient.
+    pub fn try_acquire(&self, bytes: u64) -> Option<CompactionMemoryGuard> {
+        match &self.inner {
+            None => Some(CompactionMemoryGuard::unlimited()),
+            Some(inner) => {
+                let permits = bytes_to_permits(bytes);
+                if permits == 0 {
+                    return Some(CompactionMemoryGuard::unlimited());
+                }
+
+                match inner.semaphore.try_acquire_many(permits) {
+                    Ok(permit) => {
+                        permit.forget();
+                        inner.on_acquire();
+                        Some(CompactionMemoryGuard::limited(inner.clone(), permits))
+                    }
+                    Err(TryAcquireError::NoPermits) | Err(TryAcquireError::Closed) => {
+                        COMPACTION_MEMORY_REJECTED
+                            .with_label_values(&["try_acquire"])
+                            .inc();
+                        None
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl Inner {
+    fn on_acquire(&self) {
+        self.update_in_use_metric();
+    }
+
+    fn release(&self, permits: u32) {
+        if permits == 0 {
+            return;
+        }
+        self.semaphore.add_permits(permits as usize);
+        self.update_in_use_metric();
+    }
+
+    fn used_permits(&self) -> u32 {
+        self.limit_permits
+            .saturating_sub(self.available_permits_clamped())
+    }
+
+    fn available_permits_clamped(&self) -> u32 {
+        // Clamp to limit_permits to ensure we never report more available permits
+        // than our configured limit, even if semaphore state becomes inconsistent.
+        self.semaphore
+            .available_permits()
+            .min(self.limit_permits as usize) as u32
+    }
+
+    fn update_in_use_metric(&self) {
+        let bytes = permits_to_bytes(self.used_permits());
+        COMPACTION_MEMORY_IN_USE.set(bytes as i64);
+    }
+}
+
+/// Guard representing a slice of reserved compaction memory.
+///
+/// NOTE: `AdditionalMemoryGuard` must not outlive this guard. Dropping this guard
+/// will release both base and additional permits; any surviving additional guard
+/// would then observe zero permits and run without actually holding memory.
+pub struct CompactionMemoryGuard {
+    state: GuardState,
+}
+
+enum GuardState {
+    Unlimited,
+    Limited {
+        inner: Arc<Inner>,
+        /// Base quota allocated at task startup.
+        base_permits: u32,
+        /// Additional memory allocated during task execution.
+        additional_permits: Arc<AtomicI64>,
+    },
+}
```

Comment:
Can we use OwnedSemaphorePermit to implement the guard? The CompactionMemoryGuard/AdditionalMemoryGuard can hold an OwnedSemaphorePermit.

You can also merge two permits so you may not have to define a new type for additional guard.
https://docs.rs/tokio/latest/tokio/sync/struct.OwnedSemaphorePermit.html#method.merge